### PR TITLE
vmsa: Updating for kernel debug_swap

### PR DESCRIPTION
**Pull request by larrydewey**
The upstream host kernel patches have a new
`debug_swap` feature which modifies the initial
state each vCPU's VMSA page, which may affect
the generated measurement. This change should
address the requirements introduced by the
upstream kernel change.

Resolves: [#195](https://github.com/AMDESE/AMDSEV/issues/195)
Resolves: #30

### CODE REVIEW

**Comment by dubek** [comment]
Thanks @larrydewey . The fix itself looks fine, but raises a few questions:

1. Can you please link to the kernel commit that modified this value?
2. Unit-tests are now (obviously) broken.
3. Do we want to mandate use of kernel >= X.Y.Z (whichever version that added debug_swap upstream) when sev-snp-measure is used? If we want, this should be documented. @tylerfanelli @fitzthum What's you opinion here?

**Comment by larrydewey** [comment]
> Thanks @larrydewey . The fix itself looks fine, but raises a few questions:
>
> 1. Can you please link to the kernel commit that modified this value?

It looks like it was introduced with commit `d1f85fbe836e6` by `Alexey Kardashevskiy`. 

> 2. Unit-tests are now (obviously) broken.

Excellent point! I forgot to add a `WIP` to the title of this PR. The tests definitely need to be updated. I need to figure out exactly how we can test for this change, though.

> 3. Do we want to mandate use of kernel >= X.Y.Z (whichever version that added debug_swap upstream) when sev-snp-measure is used? If we want, this should be documented. @tylerfanelli @fitzthum What's you opinion here?

I think this would be a good approach. An alternative may be introducing logic to cover both cases? Other opinions are definitely welcome.

**Comment by larrydewey** [comment]
@dubek I need to update the tests cases on this, but do you have any thoughts/feelings on the logic? I just re-based on the latest changes, so I **_think_** this should be good to go after the tests are updated.

**Review by tylerfanelli — APPROVED** [review]
LGTM. Thanks for updating the tests.

**Comment by tylerfanelli** [comment]
> 3. Do we want to mandate use of kernel >= X.Y.Z (whichever version that added debug_swap upstream) when sev-snp-measure is used? If we want, this should be documented. @tylerfanelli @fitzthum What's you opinion here?

Agreed, especially if the feature can be turned off (which also should be documented if so).

File: sevsnpmeasure/vmsa.py
Comment (dubek) [inline sevsnpmeasure/vmsa.py:190]: let's remove this condition.
When calling from `seves_calc_launch_digest`, always pass 0x0 for the guest_features argument.
When calling from the SNP launch digest functions, pass the user-defined guest_features from the command-line.

File: README.md
Comment (dubek) [inline README.md:70]: I usually put the usage section in this README by pasting the output of

    COLUMNS=100 python3 ./sev-snp-measure.py --help

but if you try that, you get something like:

```
  --initrd PATH         Initrd file to calculate hash from (use with --kernel)
  --append CMDLINE      Kernel command line to calculate hash from (use with --kernel)
  --guest-features VALUE
                        Hex representation of the guest kernel features expected to be included (Defaults to 0x21). | BIT FIELD |
                        Description | |:---------:|:------------| | 0 | SNPActive | | 1 | vTOM | | 2 | ReflectVC | | 3 |
                        RestrictedInjection | | 4 | AlternateInjection | | 5 | DebugSwap | | 6 | PreventHostIBS | | 7 | BTBIsolation
                        | | 8 | VmplSSS | | 9 | SecureTSC | | 10 | VmgexitParameter | | 11 | Reserved, SBZ | | 12 | IbsVirtualization
                        | | 13 | Reserved, SBZ | | 14 | VmsaRegProt | | 15 | SmtProtection | | 63:16 | Reserved, SBZ | Ex. **
                        --guest-features 0x21 -> DebugSwap + SNPActive ** --guest-features 0x211 -> SecureTSC + DebugSwap + SNPActive
  --output-format {hex,base64}
                        Measurement output format
```

not exactly the pretty table we want...

I suggest adding a separate section in the README which includes this table ("SNP guest features field values"). In the argument description have something shorter like (untested):

```python3
 parser.add_argument('--guest-features', metavar='VALUE', type=int,
                         default=0x21,
                        help="Hex value of the guest kernel features expected to be included (defaults to 0x21); see README.md for possible values.")
```

In the new README section we can also description the evolution (until kernel 6.6 the value was 0x1 but then it was changed to 0x21 (see commit WXYZ)).

File: sevsnpmeasure/cli.py
Comment (dubek) [inline sevsnpmeasure/cli.py:15]: no, I'll do that separately. Please remove.
I might release a version before merging this and then after.

**Review by dubek — COMMENTED** [review]
Thanks @larrydewey and @tylerfanelli . See my comments below.

**Review by dubek — APPROVED** [review]
Thanks @larrydewey , this looks good to me.

I think I'll release a new 0.0.8 today before merging this, and then release 0.0.9 with this change. This way, if users want the previous behaviour (and don't want to add `--guest-features=0x1`) they can `pip install sev-snp-measure==0.08`

@larrydewey @tylerfanelli Does that sounds fine to you?

**Review by tylerfanelli — APPROVED** [review]

**Comment by tylerfanelli** [comment]
@dubek Sounds good to me. I won't merge right now, as to give you the opportunity to release 0.0.8. Thanks!

**Comment by dubek** [comment]
Merged. Thanks @larrydewey . I'll add a notice to the changelog that this is a breaking change (for kernel < 6.6).

**Pull request merged**
